ipld/read.go

```diff
@@ -20,188 +18,130 @@ import (
-	flattened := sc.flatten()
-	// repair the square
-	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(edsWidth) / 2)
-	return rsmt2d.RepairExtendedDataSquare(rowRoots, colRoots, flattened, codec, tree.Constructor)
-}
 
-// uniqueRandNumbers generates count unique random numbers with a max of max
-func uniqueRandNumbers(count, max int) []uint32 {
-	if count > max {
-		panic(fmt.Sprintf("cannot create %d unique samples from a max of %d", count, max))
-	}
-	samples := make(map[uint32]struct{}, count)
-	for i := 0; i < count; {
-		// nolint:gosec // G404: Use of weak random number generator
-		sample := uint32(rand.Intn(max))
-		if _, has := samples[sample]; has {
-			continue
-		}
-		samples[sample] = struct{}{}
-		i++
-	}
-	out := make([]uint32, count)
-	counter := 0
-	for s := range samples {
-		out[counter] = s
-		counter++
-	}
-	return out
-}
-
-type index struct {
-	row uint32
-	col uint32
-}
-
-type indexedShare struct {
-	data []byte
-	index
-}
-
-// shareCounter is a thread safe tallying mechanism for share retrieval
-type shareCounter struct {
-	// all shares
-	shares map[index][]byte
-	// number of shares successfully collected
-	counter uint32
-	// the width of the extended data square
-	edsWidth uint32
-	// the minimum shares needed to repair the extended data square
-	minSharesNeeded uint32
-
-	shareChan chan indexedShare
-	ctx       context.Context
-	cancel    context.CancelFunc
-	// any errors encountered when attempting to retrieve shares
-	errc chan error
+	tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(edsWidth) / 2)
+	return rsmt2d.RepairExtendedDataSquare(rowRoots, colRoots, shares.data, codec, tree.Constructor)
 }
 
-func newshareCounter(parentCtx context.Context, edsWidth uint32) *shareCounter {
-	ctx, cancel := context.WithCancel(parentCtx)
-
-	// calculate the min number of shares needed to repair the square
-	minSharesNeeded := edsWidth * edsWidth / 4
+// fillBlockWithData fetches 1/4 of shares for the given root
+func fillBlockWithData(
+	ctx context.Context,
+	data [][]byte, dag ipld.NodeGetter,
+	isRow bool,
+	f *totalShares,
+) {
+	for i := 0; i < len(data)/2; i++ {
+		go func(i int) {
+			rootHash, err := plugin.CidFromNamespacedSha256(data[i])
+			if err != nil {
+				f.err <- err
+				return
+			}
+			subtreeRootHash, err := getHalfTreeNodeHash(ctx, rootHash, dag, true)
+			if err != nil {
+				f.err <- err
+				return
+			}
 
-	return &shareCounter{
-		shares:          make(map[index][]byte),
-		edsWidth:        edsWidth,
-		minSharesNeeded: minSharesNeeded,
-		shareChan:       make(chan indexedShare, 512),
-		errc:            make(chan error, 1),
-		ctx:             ctx,
-		cancel:          cancel,
+			leafs, err := GetLeafs(ctx, dag, *subtreeRootHash)
+			if err != nil {
+				f.err <- err
+				return
+			}
+			for leafIdx, leaf := range leafs {
+				shareData := leaf.RawData()[1:]
+				// it's not needed to store data for cols
+				// as we are fetching data from the same share for rows and cols
```

Comment:
Can you also add a document as to why we still fetch cols even though we're not storing them in the data slice? cc @Wondertan

Reply:
As per multiple discussionss, no need to fetch cols, but only as retry if fails for rows